### ISSUE TITLE
Support MacOs/Unix devices via NET5 ?

### ISSUE THREAD

**Issue by jkone27** [melanchall/drywetmidi#132]
would be nice if the library also supported devices on MacOs/unix/linux via .net5

**Comment by melanchall** [comment]
Hi,

Right now I'm working on macOS support for Devices API. The support will be in the next release.

As for *nix I can't give any dates when it will be supported.

Please note that .NET5 don't automatically bring devices support on macOS/*nix. It is still platform-specific and need to be implemented manually for each operating system.

**Comment by david-a-angel** [comment]
Hi melanchall,
Any idea when this release is expected to come out? I am also looking forward to having Devices API on my macOS.
Thank you!

**Comment by melanchall** [comment]
Hi @david-a-angel,

I think I'll release the new version in the middle of October. But the new API is already implemented, so I can build prerelease NuGet package for you. Would it be useful for you?

**Comment by david-a-angel** [comment]
Absolutely! And also send me a link where I can buy you a coffee or beer...or Both!!

**Comment by melanchall** [comment]
Version [6.0.0-prerelease3](https://www.nuget.org/packages/Melanchall.DryWetMidi/6.0.0-prerelease3) is published. For macOS additional features available like `VirtualDevice` and `DevicesWatcher`.

You can click on **Support** button and choose a way to support me if you want :-)

**Comment by david-a-angel** [comment]
Still Throwing a System.DllNotFoundException
Below is the simple application that is throwing Exception.

namespace DryWetMidiDebug
{
    class MainClass
    {
        public static void Main(string[] args)
        {
            var devices = InputDevice.GetAll().ToList();

            Console.WriteLine("Hello World!");
        }
    }
}

Below is the Trace.

System.DllNotFoundException: Melanchall_DryWetMidi_Native64 assembly:<unknown assembly> type:<unknown type> member:(null)
  at at (wrapper managed-to-native) Melanchall.DryWetMidi.Devices.CommonApi64.GetApiType()
  at Melanchall.DryWetMidi.Devices.CommonApi64.Api_GetApiType () [0x00000] in <51c394730c114f5583e91f6f4842b7bd>:0
  at Melanchall.DryWetMidi.Devices.MidiDevicesSession.GetSessionHandle () [0x00047] in <51c394730c114f5583e91f6f4842b7bd>:0
  at Melanchall.DryWetMidi.Devices.MidiDevice.EnsureSessionIsCreated () [0x00000] in <51c394730c114f5583e91f6f4842b7bd>:0
  at Melanchall.DryWetMidi.Devices.InputDevice+<GetAll>d__36.MoveNext () [0x00017] in <51c394730c114f5583e91f6f4842b7bd>:0
  at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs:1108
  at System.Collections.Generic.List`1[T]..ctor (System.Collections.Generic.IEnumerable`1[T] collection) [0x00062] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs:87
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0000e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/ToCollection.cs:30
  at DryWetMidiDebug.MainClass.Main (System.String[] args) [0x00001] in /Users/davidangel/Development/DryWetMidiDebug/DryWetMidiDebug/Program.cs:11

**Comment by melanchall** [comment]
@david-a-angel Thanks for the log.

All my unit tests are passed on macOS, so obviously our environments are different. All my tests are done with `dotnet` commands (like `dotnet run`, `dotnet build`, etc), but in your log I see mentions of **mono**, so I suppose you use Mono for cross-platform .NET developement. From the [documentation on DllImport for macOS](https://www.mono-project.com/docs/advanced/pinvoke/#macos-framework-and-dylib-search-path) for Mono:

> The Framework and library search path is:
> 
> 1. A colon-separated list of directories in the user’s DYLD_FRAMEWORK_PATH environment variable.
> 2. A colon-separated list of directories in the user’s DYLD_LIBRARY_PATH environment variable.
> 3. A colon-separated list of directories in the user’s DYLD_FALLBACK_FRAMEWORK_PATH environment variable, which defaults to the directories:  
>    * ~/Library/Frameworks
>    * /Library/Frameworks
>    * /Network/Library/Frameworks
>    * /System/Library/Frameworks
> 4. A colon-separated list of directories in the user’s DYLD_FALLBACK_LIBRARY_PATH environment variable, which defaults to the directories:
>    * ~/lib
>    * /usr/local/lib
>    * /lib
>    * /usr/lib

So it looks like Mono unable to search for libs in current directory which is sad :( I should investigate integration with Mono.

I have some questions:

1. Can you use .NET Core / .NET instead of Mono?
2. Can you please share me content of the `/bin` directory of your application so we can see whether native libs are copied there or not?
3. What if you put `Melanchall_DryWetMidi_Native64.dylib` to some directory from the list from the Mono documentation above? Will the app work?

Thanks

**Comment by david-a-angel** [comment]
It is a basic console application on Visual Studio Mac. 
Files Inside bin directory are below separated by ' | '.
DryWetMidiDebug.exe  |   DryWetMidiDebug.pdb  |  Melanchall_DryWetMidi_Native32.dll  |  Melanchall_DryWetMidi_Native64.dll  |  Melanchall_DryWetMidi_Native64.dylib  |  Melanchall.DryWetMidi.dll  |  Melanchall.DryWetMidi.xml 

Part of the Library is working. For instance,
var file = MidiFile.Read("K545-1.mid");
works as expected. Returning a MidiFile object as expected. It is just the Devices namespace that is having problems.

**Comment by melanchall** [comment]
@david-a-angel Thanks for the info. At least, binaries are in place. To investigate further I need following info:

1. Can you please send me entire folder with your solution?
2. What version of your macOS?
2. What versoin of the VS for Mac you use?
3. What version of .NET installed on your machine? (run `dotnet --version` in Terminal)
4. What if you run your app via `dotnet run --project path/to/your/project.csproj -c Release` in Terminal?

**Comment by david-a-angel** [comment]
For
1. [DryWetMidiDebug.zip](https://github.com/melanchall/drywetmidi/files/7246343/DryWetMidiDebug.zip)
2. My macOS version: macOS Big Sur version 11.4
3. My VS is: Visual Studio for Mac Community Version 8.10.8
4. .Net version: 5.0.4
5.  The Project has nothing to run. It only is checking if the InputDevice.GetAll() is returning with anything.

**Comment by melanchall** [comment]
Thanks for the info.

2. I've checked on 10.14 Mojave with the latest VS for Mac and all the Devices API works. I'll check your solution with Big Sur and let you know.
5. Just print all devices with `Console.WriteLine` and run the project with the command I've provided.

**Comment by david-a-angel** [comment]
Here is the results for the dotnet run in Terminal.

The build failed. Fix the build errors and run again.
Davids-MacBook-Pro:~ davidangel$ dotnet run --project /Development/DryWetMidiDebug/DryWetMidiDebug/DryWetMidiDebug.csproj -c Release
MSBUILD : error MSB1001: Unknown switch.
Switch: /Development/DryWetMidiDebug/DryWetMidiDebug/DryWetMidiDebug.csproj

**Comment by melanchall** [comment]
Thanks for the archive. Now I see what's going on. You've created solution of the wrong type. It looks like you've created solution on Windows targeting .NET Framework and then want to run it on macOS. Thta's wrong. .NET Framework doesn't work on systems other than Windows.

Just create solution from scratch on macOS in Visual Studio for Mac. Or you can create it on Windows, BUT you need to target **.NET Core or .NET 5**.

So your bug is not related with DryWetMIDI. .NET Framework projects can be run on Windows only. .NET Core and .NET 5 ones – on Windows, macOS and Linux.

Let me know, whether fresh macOS project works or not. If not – send me solution again.

**Comment by david-a-angel** [comment]
Interesting. Whenever I create a console app on Visual Studio for Mac it automatically creates a .Net Framework project and there is no way to specify otherwise. I had to go to my PC and create a .NET Core console app then send it to my mac. Now it appears to be working. I haven't done much with the input / output devices yet but the call to GetAll() is not throwing an exception anymore.
Thanks for all your help!
